tsl/src/continuous_aggs/planner.c

```diff
@@ -0,0 +1,316 @@
+					lappend(context->watermark_parent_functions, context->parent_to_timestamp_func);
+			}
+			else
+			{
+				if (linitial(context->parent_coalesce_expr->args) != node)
+				{
+					context->valid_query = false;
+					return false;
+				}
+
+				context->watermark_parent_functions =
+					lappend(context->watermark_parent_functions, context->parent_coalesce_expr);
+			}
+		}
+
+		/* Capture the timestamp conversion function */
+		if (list_member_oid(context->to_timestamp_func_oids, funcExpr->funcid))
+		{
+			FuncExpr *old_func_expr = context->parent_to_timestamp_func;
+			context->parent_to_timestamp_func = funcExpr;
+			bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+			context->parent_to_timestamp_func = old_func_expr;
+
+			return result;
+		}
+	}
+	else if (IsA(node, Query))
+	{
+		/* Recurse into subselects */
+		Query *query = castNode(Query, node);
+		return query_tree_walker(query,
+								 constify_cagg_watermark_walker,
+								 context,
+								 QTW_EXAMINE_RTES_BEFORE);
+	}
+	else if (IsA(node, CoalesceExpr))
+	{
+		/* Capture the CoalesceExpr */
+		CoalesceExpr *parent_coalesce_expr = context->parent_coalesce_expr;
+		context->parent_coalesce_expr = castNode(CoalesceExpr, node);
+		bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+		context->parent_coalesce_expr = parent_coalesce_expr;
+
+		return result;
+	}
+	else if (IsA(node, RangeTblEntry))
+	{
+		/* Collect the Oid of the used range tables */
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+
+		if (rte->rtekind == RTE_RELATION)
+		{
+			context->relids = list_append_unique_oid(context->relids, rte->relid);
+		}
+
+		/* allow range_table_walker to continue */
+		return false;
+	}
+
+	return expression_tree_walker(node, constify_cagg_watermark_walker, context);
+}
+
+/*
+ * To avoid overhead by traversing the query tree, we perform a check before to determine if the
+ * given query could be a real-time CAgg query. So, we search for a SELECT over two subqueries.
+ */
+static bool pg_nodiscard
+could_be_realtime_cagg_query(Query *parse)
+{
+	if (parse->commandType != CMD_SELECT)
+		return false;
+
+	if (parse->hasAggs || parse->hasWindowFuncs || parse->hasTargetSRFs)
+		return false;
+
+	return true;
+}
+
+/*
+ * The entry of the watermark HTAB.
+ */
+typedef struct WatermarkConstEntry
+{
+	int32 key;
+	Const *watermark_constant;
+} WatermarkConstEntry;
+
+/* The query can contain multiple watermarks (i.e., two hierarchal real-time CAggs)
+ * We maintain a hash map (hypertable id -> constant) to ensure we use the same constant
+ * for the same watermark across the while query.
+ */
+static HTAB *pg_nodiscard
+init_watermark_map()
+{
+	struct HASHCTL hctl = {
+		.keysize = sizeof(int32),
+		.entrysize = sizeof(WatermarkConstEntry),
+		.hcxt = CurrentMemoryContext,
+	};
+
+	return hash_create("Watermark const values", 2, &hctl, HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
```

Comment:
Maybe we can increase a bit the maximum number of elements of this hash table to four (4) so we'll be able to map hierarchical caggs until four levels.

Reply:
Good point, changed.